File: gui/haskell/src/GUI.hs
{-# LANGUAGE OverloadedStrings #-}
module GUI (createGUI) where

import Data.Maybe (fromJust)
import Data.GI.Base.ManagedPtr (unsafeCastTo)
import qualified GI.Gtk.Functions as GI (main, init)
import GI.Gtk (widgetShowAll, mainQuit, onButtonClicked, Button(..), Window(..), Label(..), builderGetObject, builderAddFromFile, builderNew, labelSetText)
import Network.Wreq
import Control.Lens
import Data.Aeson.Lens (_String, key)
import Data.Text.Encoding (decodeUtf8)


getVersion response = response ^. responseBody . key "version" . _String

setVersion label = do
  response <- get "http://web:8000/version"
  labelSetText label $ getVersion response

createGUI = do
  GI.init Nothing
  builder <- builderNew
  builderAddFromFile builder "../interface.glade"

  window <- builderGetObject builder "window" >>= unsafeCastTo Window . fromJust
  close <- builderGetObject builder "close" >>= unsafeCastTo Button . fromJust
  appVersion <- builderGetObject builder "app-version" >>= unsafeCastTo Label . fromJust

  refresh <- builderGetObject builder "refresh" >>= unsafeCastTo Button . fromJust
  serverVersion <- builderGetObject builder "server-version" >>= unsafeCastTo Label . fromJust


  labelSetText appVersion "Haskell"
  onButtonClicked close mainQuit
  onButtonClicked refresh $ setVersion serverVersion

  widgetShowAll window
  GI.main

File: gui/python/gui.py
#!/usr/bin/python3

import gi
import requests

gi.require_version('Gtk', '3.0')


from gi.repository import Gtk

builder = Gtk.Builder()
builder.add_from_file("interface.glade")

window = builder.get_object("window")

app_version = builder.get_object("app-version")
server_version = builder.get_object("server-version")
refresh = builder.get_object("refresh")
close = builder.get_object("close")

app_version.set_text("Python")


def update_server_version(button):
    version = requests.get('http://web:8000/version/').json()['version']
    server_version.set_text(str(version))


close.connect("clicked", lambda _: window.destroy())
refresh.connect("clicked", update_server_version)

window.show_all()
Gtk.main()
